beacon_node/http_api/src/publish_blocks.rs

```diff
@@ -119,130 +144,252 @@ pub async fn publish_block<T: BeaconChainTypes, B: IntoGossipVerifiedBlockConten
+    };
 
-    let publish_fn = move || match validation_level {
-        BroadcastValidation::Gossip => Ok(()),
-        BroadcastValidation::Consensus => publish_block(
-            block_clone,
-            blobs_opt,
-            sender_clone,
-            log_clone,
-            seen_timestamp,
-        ),
-        BroadcastValidation::ConsensusAndEquivocation => {
-            check_slashable(
-                &chain_clone,
-                &blobs_opt,
+    for blob in gossip_verified_blobs.into_iter().flatten() {
+        // Importing the blobs could trigger block import and network publication in the case
+        // where the block was already seen on gossip.
+        if let Err(e) = Box::pin(chain.process_gossip_blob(blob, &publish_fn)).await {
+            let msg = format!("Invalid blob: {e}");
+            return if let BroadcastValidation::Gossip = validation_level {
+                Err(warp_utils::reject::broadcast_without_import(msg))
+            } else {
+                error!(
+                    log,
+                    "Invalid blob provided to HTTP API";
+                    "reason" => &msg
+                );
+                Err(warp_utils::reject::custom_bad_request(msg))
+            };
+        }
+    }
+
+    match gossip_verified_block_result {
+        Ok(gossip_verified_block) => {
+            let import_result = Box::pin(chain.process_block(
                 block_root,
-                &block_clone,
-                &log_clone,
-            )?;
-            publish_block(
-                block_clone,
-                blobs_opt,
-                sender_clone,
-                log_clone,
+                gossip_verified_block,
+                NotifyExecutionLayer::Yes,
+                BlockImportSource::HttpApi,
+                publish_fn,
+            ))
+            .await;
+            post_block_import_logging_and_response(
+                import_result,
+                validation_level,
+                block,
+                is_locally_built_block,
                 seen_timestamp,
+                &chain,
+                &log,
             )
+            .await
         }
-    };
-
-    if let Some(gossip_verified_blobs) = gossip_verified_blobs {
-        for blob in gossip_verified_blobs {
-            if let Err(e) = Box::pin(chain.process_gossip_blob(blob)).await {
-                let msg = format!("Invalid blob: {e}");
-                return if let BroadcastValidation::Gossip = validation_level {
-                    Err(warp_utils::reject::broadcast_without_import(msg))
-                } else {
-                    error!(
-                        log,
-                        "Invalid blob provided to HTTP API";
-                        "reason" => &msg
-                    );
-                    Err(warp_utils::reject::custom_bad_request(msg))
-                };
+        Err(BlockError::DuplicateFullyImported(root)) => {
+            if published.load(Ordering::SeqCst) {
```

Comment:
This is always false for the `BroadcastValidation::Gossip` case, even when blobs are published I think?

Reply:
Never mind, I misunderstood the code, and there's [a test case](https://github.com/sigp/lighthouse/blob/4f3cf2c7292a127c76cc7c81063fcb408fc7a1d7/beacon_node/http_api/tests/broadcast_validation_tests.rs#L1407) that covers this scenario!

Reply:
Trying to understand the `published` logic here - in what scenario would this be `false` here?

When `BroadcastValidation::Gossip` == true, we hit this code path if
- gossip verification for block fails with `DuplicateFullyImported`
- AND none of the blob import fails
- `published` is always true in this scenario

When validation level is not gossip, `published` remains `false` if `publish_block_p2p` fails, or if if we get a `BlockError::DuplicateFullyImported` when processing blobs - in both cases we return a `custom_bad_request` rejection I think?

Reply:
Oh yes you're right. I think I intended to _not_ return an error early here if the blobs are duplicates:

https://github.com/sigp/lighthouse/blob/27113b28960fff9462dc682dd2b09b93025de63e/beacon_node/http_api/src/publish_blocks.rs#L275-L276

The `false` case is meant to be the one where block & blobs were fully known and already imported.

Reply:
I'll try writing a new test that fails and then tweaking the logic to make it pass

Reply:
Ok the test I wrote passed, and also hit this branch! :tada:

See commit 52495eeef31289cb05e776dfeb3c5f3cb183ecd3

The reason it doesn't fail at the `process_gossip_blob` step is because the duplicate blobs have already been filtered out (converted to `None`) when we construct the gossip-verified blobs